returnn/tf/layers/basic.py

```diff
@@ -1341,6 +1341,17 @@ def __init__(self, position, axis, **kwargs):
     # (BatchAxes.., InputAxesBeforeGatherAxis, PositionAxes.., InputAxesAfterGatherAxis..)
     self.output.placeholder = tf.gather(params=params, indices=indices, axis=gather_axis, batch_dims=batch_dims)
 
+    if input_data.dim_tags[old_gather_axis].is_batch_dim():
+      for axis in self.output.size_placeholder:
+        new_size = tf.gather(params=self.output.size_placeholder[axis], indices=position_data.placeholder)
```

Comment:
You assume that `position_data` is of shape [new-batch]?

Reply:
Right, in the case I have in mind yes. But for the failing test case, this is different and we need to take this into account.

Reply:
What is it in that case?

Reply:
There it's of shape [B,T,F], however, in the input B and T are packed
```
>>> input_data                                                                                                                                                                                                                  
Data{'flat_output', [B&Packed{'time'},F|F'feature'(5)]}
>>> self.output                                                                                                                                                                                                                 
Data{'output_output', [B,T|'time'[B],'other-spatial'(7),F|F'feature'(5)]}
>>> position_data                                                                                                                                                                                                               
Data{'indices_flat_output', [B,T|'time'[B],F|'other-spatial'(7)], dtype='int32'}
```

Reply:
Which test case is that? The one you added? `test_rand_indices`?
Why is `position_data` of this shape? As described, it should have some new-batch dim in it, right? Or basically just the shape [new-batch]? When you gather into the batch dim. It definitely should not have the old batch dim in its shape.

Reply:
This is from `test_rand_indices` which is an existing test case. These are the layer templates
```
layer /'data': [B,T|'time'[B],F|F'feature'(5)] float32
layer /'flat': [B&Packed{'time'},F|F'feature'(5)] float32
layer /'length_flat': [] int32
layer /'indices_flat': [B,T|'time'[B],F|'other-spatial'(7)] int32
layer /'output': [B,T|'time'[B],'other-spatial'(7),F|F'feature'(5)] float32
```
It gathers the flattened input with random indices of shape B,T,F'. That's how the test was written. So it's gathering on the flattened B&T dim, not a regular batch dim

Reply:
This is the net dict:
```
    sz = (batch_dim, time_dim, SpatialDim("other-spatial", 7))
    net.construct_from_dict({
      "flat": {"class": "flatten_batch", "from": "data"},
      "length_flat": {"class": "length", "from": "flat", "axis": batch_dim},
      "indices_flat": {"class": "rand_int", "shape": sz, "minval": 0, "maxval": "length_flat", "seed": 42},
      "output": {"class": "gather", "from": "flat", "axis": batch_dim, "position": "indices_flat"},
    })
```

Reply:
Ah yes, sorry, yes, this is a valid case, the shape of `position_data` makes sense here.

What is the output shape of `tf.gather` here, i.e. of `new_size`?

Operating (assigning) `size_placeholder` is problematic here, as this always assumes shape [B]. But I think your `new_size` has actually the same shape as `position_data`.

What is `self.output.size_placeholder[axis]` actually before? Or rather the `dyn_size_ext` of the dim tags of `input_data`?

Reply:
Yes, `new_size` has the same shape as `position_data`.
```
>>> new_size                                                                                                                                                                                                                    
<tf.Tensor 'output/GatherV2_1:0' shape=(?, ?, 7) dtype=int32>
>>> position_data.placeholder                                                                                                                                                                                                   
<tf.Tensor 'indices_flat/random_uniform:0' shape=(?, ?, 7) dtype=int32>
>>> position_data                                                                                                                                                                                                               
Data{'indices_flat_output', [B,T|'time'[B],F|'other-spatial'(7)], dtype='int32'}
```

Reply:
`self.output.size_placeholder[axis]` is of shape (B,) before.
```
>>> self.output.size_placeholder[axis]                                                                                                                                                                                          
<tf.Tensor 'extern_data/placeholders/data/data_dim0_size:0' shape=(?,) dtype=int32>
```
`dyn_size_ext` is not set on the dim tags in `input_data`.

Reply:
I wonder a bit: All the output dim tags here are already defined ([B,T|'time'[B],'other-spatial'(7),F|F'feature'(5)]). It would be wrong to overwrite the `dyn_size_ext` of any of them. So either the output dim tags are wrong (you actually want to have another time dim here?) or nothing needs to be done for this case.

Reply:
> `dyn_size_ext` is not set on the dim tags in `input_data`.

Yes because `input_data` does not have any dim tags which need that. But you are speaking about the dim tags of `self.output` here? But anyway, see my previous comment.

Reply:
> I wonder a bit: All the output dim tags here are already defined ([B,T|'time'[B],'other-spatial'(7),F|F'feature'(5)]). It would be wrong to overwrite the dyn_size_ext of any of them. So either the output dim tags are wrong (you actually want to have another time dim here?) or nothing needs to be done for this case.

Yes, I also think `dyn_size_ext` should not be changed. Regarding the different time dim I'm not sure, that's what I meant initially when saying I don't know what the desired behavior is in this case. If you don't have a clear opinion on this, I'd vote for not changing anything.

Reply:
Well, this should not be a question on opinion, but what makes sense. There should be a clear answer. Or rather, the ambiguity just comes from our lack of really understanding it, I think. But actually, if I understand it right, it's actually quite clear: The axis where the gather operates on is replaced by all the axes of `position_data`. The axes of `position_data` already exist, and all have already defined dim tags (incl dyn size), so nothing needs to be done about them, no new dyn size needs to be calculated. Maybe we still miss something in this understanding, but with this understanding, we should just leave it as it is.

Reply:
Actually, ok, one case where it needs a change: there is really a new batch dim. So some batch dim gets replaced by a new batch dim.

Reply:
So, basically, what needs to be done:

You need to check the case that you gather into the batch dim (as you do here):

Then you need to iterate over the output dim tags (similar as you do here).

For each, you need to check its dyn_size_ext, whether it contains the batch dim which you are going to replace/gather here. In your example here, it is `B&Packed{'time'}`, but this dim is not in any of the output dim tags, so nothing needs to be done. If it would be there, then you actually would need the same gather logic as you do for `output` itself.

Reply:
How do I check this? In the case with `B&Packed{'time'}`, I still get
```
>>> input_data.dim_tags[0] == self.output.dim_tags[0]                                                                                                                                                                           
True
```
Even though the input has the packed one and the output the global batch. See https://github.com/rwth-i6/returnn/blob/master/returnn/tf/util/data.py#L766 Do we need to extend that logic for this case?

Reply:
Yea, that's the problem I mentioned elsewhere, that dim tag equality is handled differently for the batch dim...

Here, in this case, we need different logic for the equality check of the batch dim. For example, you can compare `batch_dim.batch` or `data.batch`.

Reply:
In the output, you anyway already need to assign `output.batch`. So you can compare `input_data.batch != output.batch`. For each output dim tag, you should call `get_for_batch_ctx`. And then in `__init__`, the `dyn_size_ext` should be unknown for the case that there is a new batch dim, and if so, then you need the gather logic also on it. You get the old original dyn size again via `get_for_batch_ctx`.

Reply:
`*.batch` is `None` here.

Reply:
It should not be `None`. For `input_data`? Or `output`? Or `position_data`? For `output`, you need to assign it.

Reply:
Right now, it's `None` for all.

Reply:
I see that I need to assign it for `output`. But it should come from `position_data`, right?

Reply:
Why is it `None` for `position_data`? I don't mean in the test case, I mean in the real case which motivated this test case. In the real case, you would not have such `InternalLayer`.

Reply:
It should never be done if the data has a batch dim, unless sth is wrong. In case of the test case, then the test case is buggy.

Reply:
Right, this is about the test case. However, in the case that I'm interested in, still `input_data.batch == position_data.batch` is `True`. This is probably because I'm using an `EvalLayer` to get the batch indices from a 0/1 vector with shape (B,) and that `EvalLayer` does not set the output correctly. Then we would need a layer which does that correctly, right?

Reply:
An `EvalLayer` should never change the shape. If it does, and you are not very careful in setting the output data, then yes, this is a bug in your config.